sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,456 @@
+
+Assert.DoesNotThrowAsync(async () => await sendFailed(args));
+
+// The expected outcome of any application-specific complex processing can be asserted here.
+```
+
+## Consuming events and reading properties from an `EventHubConsumerClient`
+
+When testing code that is dependent on the `EventHubConsumerClient`, a developer may want to create a known set of expected events and contexts to verify that their application consumes them correctly.  Tests can mock `EventData` instances coming from the broker using the `EventHubModelFactory`. These can be used within the consumer client mock to mock returns from `ReadEventsFromPartitionAsync`, which for most applications is the key interaction to test.
+
+This snippet also demonstrates how to mock `LastEnqueuedEventProperties` using the `EventHubsModelFactory`.
+
+```C# Snippet:EventHubs_Sample11_MockingConsumerClient
+Mock<EventHubConsumerClient> mockConsumer = new();
+
+// This creates a mock LastEnqueuedEventProperties using the model factory.
+
+LastEnqueuedEventProperties lastEnqueueEventProperties = EventHubsModelFactory.LastEnqueuedEventProperties(
+    lastSequenceNumber : 1234,
+    lastOffset : 234,
+    lastEnqueuedTime : DateTimeOffset.Parse("1:24 AM"),
+    lastReceivedTime : DateTimeOffset.Parse("1:26 AM"));
+
+// This creates a mock of PartitionContext using the model factory.
+
+PartitionContext partitionContext = EventHubsModelFactory.PartitionContext(
+    partitionId : "0",
+    lastEnqueuedEventProperties : lastEnqueueEventProperties);
+
+// This is a simple local method that returns an IAsyncEnumerable to use as the return for
+// ReadEventsAsync below, since IAsyncEnumerables cannot be created directly.
+
+async IAsyncEnumerable<PartitionEvent> mockReturn()
+{
+    for (int index = 0; index < 3; index++)
+    {
+        // This mocks an EventData instance using the model factory. Different arguments can mock different
+        // potential outputs from the broker.
+
+        EventData eventData = EventHubsModelFactory.EventData(
+            eventBody: new BinaryData($"Sample-Event-{ index }"),
+            systemProperties: new Dictionary<string, object>(), //arbitrary value
+            partitionKey: "sample-key",
+            sequenceNumber: 1000,
+            offset: 1500,
+            enqueuedTime: DateTimeOffset.Parse("11:36 PM"));
+
+        // This creates a mock PartitionEvent to return from the consumer client.
+
+        PartitionEvent samplePartitionEvent = new PartitionEvent(partitionContext, eventData);
+        List<PartitionEvent> partitionEventList = new List<PartitionEvent> { samplePartitionEvent };
+
+        // IAsyncEnumerable types can only be returned by async functions, use this no-op await statement to
+        // force the method to be async.
+
+        await Task.Yield();
+
+        // Yield statements allow methods to emit multiple outputs. In async methods this can be over time.
+
+        yield return samplePartitionEvent;
```

Comment:
It would be helpful to yield multiple events; there's often confusion about `yield return` being similar to `return` in that "you can only use one".